### PR TITLE
log compaction

### PR DESCRIPTION
**Pull request by MarinPostma**
# Log compaction and snapshotting

This PR enables log compaction and snapshotting.

## Motivation

The replication log follows sqlite WAL and grows indefinitely. Fortunately, it contains a lot of duplicate data, so we can compress it by getting by keeping only the most recent version of each page. This is what this PR does: whenever the replication log grows above some threshold, a new log is created, and the old log is compacted. This operation is done atomically, so the compaction can happen in the background, while we keep writing to the old log.

## Log compaction:

The log compaction is very straightforward. We iterate backwards through the replication log, and write the frames to the snapshot file. We keep track of what pages we have already seen, and ignore older version of them. When the snapshot is finished, we remove the old log file.

Notice how the frames are now in a reverse order in the snapshot: starting with the most recent, and ending with the oldest.

## Snapshoting:

Whenever a replica asks for a frame that is not present in the current log (i.e, it asks for a frame index less than the log starting frame id), then it sends the replica an error, asking it to ask for a snapshot.

The replica receives this message, and immediately asks for a snapshot, sending over the frame id that got it rejected in the first place.
The primary looks for a snapshot containing the request frame, and starts iterating through it, until it reaches a frame that is less than the requested one.

This mechanism allows us to send partial snapshot: the replica gets minimal amount of frames required to get up to speed.

The replica writes the snapshot frames to a file, and then `mmap` that file to build a chained list of pages to append to the WAL.

## Future work:

This PR was getting a bit too long, so I left out some work for followup PRs:

- Even though the snapshot significantly compresses the size of the log, a new log is created, that will lead to the creation of a new snapshot. Those snapshot will pile up and we're back at step one. The next step is to merge snapshots into bigger snapshots, and get rid of the older snapshots.
- Every query for a frame causes a read to the snapshot/log. To speed things up, let's add a MRU cache.
- Explore compression

### CODE REVIEW

File: sqld/src/replication/logger.rs [hunk 1]
```diff
@@ -442,16 +605,45 @@ impl ReplicationLogger {
...
+    }
+
+    pub async fn locate_snapshot(
+        &self,
+        start_idx: FrameId,
+    ) -> anyhow::Result<Option<SnapshotFile>> {
+        static SNAPSHOT_FILE_MATCHER: Lazy<Regex> = Lazy::new(
+            ||
+            //                                                   v---v-end idx
+            Regex::new(r#"(\w{8}-\w{4}-\w{4}-\w{4}-\w{12})-(\d*)-(\d*).snap"#).unwrap(),
```
Comment (honzasp) [inline sqld/src/replication/logger.rs:623]: Tip: you can use the `x` regex flag (`(?x)` at the beginning) to ignore whitespace in the regex, which allows you to split the regex into multiple lines and include comments:
https://docs.rs/regex/latest/regex/#grouping-and-flags

**Review by honzasp — COMMENTED** [review]
Proper review of this code would take days, so I have just some superficial comments. I haven't found any actual issues; the high-level design is reasonable, but I found the code a bit hard to follow. I think that this is caused by a combination of multiple reasons:
1. The whole approach of working on the level of WAL frames, instead of on the level of pages and versions, is a bit confusing. However, I understand the reasons that led to this design.
2. The replication code interweaves the gRPC communication mechanism, the SQLite hacks and our replication logic.
3. Sometimes the naming is a bit inconsistent; for example, the code sometimes say "frame id" and sometimes "frame index", the word "offset" often means multiple things in the same context, etc. But this is probably my personal preference, because I'm very pedantic with naming in my own code (to an almost autistic extent! :))

File: sqld/src/replication/snapshot.rs [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+pub struct SnapshotFileHeader {
+    /// id of the database
+    pub db_id: u128,
+    /// first frame in the snapshot
+    pub start_frame_id: u64,
+    /// end frame in the snapshot
+    pub end_frame_index: u64,
+    /// number of frames in the snapshot
+    pub frame_count: u64,
+    /// safe of the database after applying the snapshot
```
[the remark this answers is not in the document]
Comment (MarinPostma) [reply sqld/src/replication/snapshot.rs:28]: I'll add units 👍 
the database size is in pages, and it comes from sqlite.

File: sqld/src/replication/logger.rs (hunk 1 above)
Comment (MarinPostma) [reply sqld/src/replication/logger.rs:623]: nice

File: sqld/src/replication/snapshot.rs [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+                    if header.frame_id <= offset {
+                        None
+                    } else {
+                        Some(Ok(buf.freeze()))
+                    }
+                }
+                Err(e) => Some(Err(e.into())),
+            }
+        })
+    }
```
[the remark this answers is not in the document]
Comment (MarinPostma) [reply sqld/src/replication/snapshot.rs:74]: I'll work on that

**Comment by MarinPostma** [comment]
bors merge
